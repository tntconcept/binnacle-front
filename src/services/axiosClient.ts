import axios, { AxiosInstance } from "axios";
import { getToken } from "core/contexts/AuthContext/tokenUtils";

export const axiosClient: AxiosInstance = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
  timeout: 10000
});

axiosClient.interceptors.request.use(
  requestConfig => {
    const accessToken = getToken("access_token");
    if (accessToken) {
      requestConfig.headers.Authorization = `Bearer ${accessToken}`;
    }

    return requestConfig;
  },
  error => Promise.reject(error)
);
