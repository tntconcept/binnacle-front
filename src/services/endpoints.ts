export const AUTH_ENDPOINT = "/oauth/token";
export const USER_ENDPOINT = "/api/user";
export const ACTIVITIES_ENDPOINT = "/api/activities";
export const TIME_TRACKER_ENDPOINT = "/api/timeBalance";
export const HOLIDAYS_ENDPOINT = "/api/holidays";
export const ORGANIZATION_ENDPOINT = "/api/organization";
export const PROJECT_ENDPOINT = "/api/project";
export const PROJECT_ROLE_ENDPOINT = "/api/role";
