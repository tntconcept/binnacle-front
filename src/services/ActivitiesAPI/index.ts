import {index} from "services/HttpClient"
import {ACTIVITIES_ENDPOINT} from "services/endpoints"
import {formatDateForQuery} from "utils/DateUtils"
import {IActivity, IActivityDay, IActivityRequestDTO} from "interfaces/IActivity"
import {activityDateMapper, activityDayMapper} from "services/ActivitiesAPI/mapper"

export const getActivitiesBetweenDate = async (
  startDate: Date,
  endDate: Date
) => {
  const result = await index
    .get(ACTIVITIES_ENDPOINT, {
      searchParams: {
        startDate: formatDateForQuery(startDate),
        endDate: formatDateForQuery(endDate)
      }
    })
    .json<IActivityDay[]>();

  return result.map(activityDayMapper);
};

export const createActivity = async (
  activity: Omit<IActivityRequestDTO, "id">
) => {
  const result = await index
    .post(ACTIVITIES_ENDPOINT, {
      json: { ...activity }
    })
    .json<IActivity>();

  return activityDateMapper(result)
};

export const updateActivity = async (activity: IActivityRequestDTO) => {
  const result = await index.put(ACTIVITIES_ENDPOINT, {
    json: {...activity}
  }).json<IActivity>()
  return activityDateMapper(result)
};

export const deleteActivity = async (id: number) => {
  return await index.delete(`${ACTIVITIES_ENDPOINT}/${id}`).text()
};

export const getActivityImage = async (id: number) => {
  return await index.get(`${ACTIVITIES_ENDPOINT}/${id}/image`).text()
}
