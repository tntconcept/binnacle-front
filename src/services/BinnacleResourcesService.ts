import { endOfMonth, startOfMonth, startOfYear } from 'date-fns'
import {
  firstDayOfFirstWeekOfMonth,
  lastDayOfLastWeekOfMonth
} from 'utils/DateUtils'
import DateTime from 'services/DateTime'
import { fetchHolidaysBetweenDate } from 'api/HolidaysAPI'
import { fetchTimeBalanceBetweenDate } from 'api/TimeBalanceAPI'
import { fetchActivitiesBetweenDate } from 'api/ActivitiesAPI'
import { fetchRecentRoles } from 'api/RoleAPI'

const buildTimeBalanceKey = (month: Date) => {
  const monthNumber = ('0' + (month.getMonth() + 1).toString()).slice(-2)
  return month.getFullYear() + '-' + monthNumber + '-01'
}

class BinnacleService {
  fetchTimeBalance = async (month: Date, mode: 'by_month' | 'by_year') => {
    const promise =
      mode === 'by_month'
        ? this.fetchTimeDataByMonth(month)
        : this.fetchTimeDataByYear(month)

    return await promise
  }

  async fetchTimeDataByMonth(month: Date) {
    const startDate = startOfMonth(month)
    const endDate = endOfMonth(month)

    const data = await fetchTimeBalanceBetweenDate(startDate, endDate)

    return data[buildTimeBalanceKey(month)]
  }

  async fetchTimeDataByYear(month: Date) {
    const response = await fetchTimeBalanceBetweenDate(
      startOfYear(month),
      endOfMonth(month)
    )

    const onlySelectedYear = Object.fromEntries(
      Object.entries(response).filter(([key]) =>
        key.includes(month.getFullYear().toString())
      )
    )

    const totalTimeStats = Object.values(onlySelectedYear).reduce(
      (prevValue, currentValue) => ({
        timeWorked: prevValue.timeWorked + currentValue.timeWorked,
        timeToWork: prevValue.timeToWork + currentValue.timeToWork,
        timeDifference: prevValue.timeDifference + currentValue.timeDifference
      })
    )

    return totalTimeStats
  }

  async fetchHolidays(month: Date) {
    const firstDayOfFirstWeek = firstDayOfFirstWeekOfMonth(month)
    const lastDayOfLastWeek = lastDayOfLastWeekOfMonth(month)

    return await fetchHolidaysBetweenDate(firstDayOfFirstWeek, lastDayOfLastWeek)
  }

  async fetchActivities(month: Date) {
    const firstDayOfFirstWeek = firstDayOfFirstWeekOfMonth(month)
    const lastDayOfLastWeek = lastDayOfLastWeekOfMonth(month)

    const isThisMonthOrPrevious =
      DateTime.isThisMonth(month) ||
      DateTime.isSameMonth(month, DateTime.subMonths(DateTime.now(), 1))

    const [activities, recentRoles] = await Promise.all([
      fetchActivitiesBetweenDate(firstDayOfFirstWeek, lastDayOfLastWeek),
      isThisMonthOrPrevious ? fetchRecentRoles() : undefined
    ])

    return {
      activities,
      recentRoles
    }
  }
}

export const BinnacleResourcesService = new BinnacleService()
