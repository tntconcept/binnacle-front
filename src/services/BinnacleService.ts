import React from "react"
import {firstDayOfFirstWeekOfMonth, lastDayOfLastWeekOfMonth} from "utils/DateUtils"
import {endOfMonth, isSameMonth, isThisMonth, startOfMonth, startOfYear, subDays} from "date-fns"
import {getActivitiesBetweenDate} from "api/ActivitiesAPI"
import {getHolidaysBetweenDate} from "api/HolidaysAPI"
import {getTimeBalanceBetweenDate} from "api/TimeBalanceAPI"
import {BinnacleActions, TBinnacleActions} from "core/contexts/BinnacleContext/BinnacleActions"
import {getRecentRoles} from "api/RoleAPI"

const buildTimeBalanceKey = (month: Date) => {
  const monthNumber = ('0' + (month.getMonth() + 1).toString()).slice(-2)
  return month.getFullYear() + "-" + monthNumber + "-01"
}

export const fetchBinnacleData = async (
  month: Date,
  isTimeCalculatedByYear: boolean,
  dispatch: React.Dispatch<TBinnacleActions>
) => {
  const firstDayOfFirstWeek = firstDayOfFirstWeekOfMonth(month);
  const lastDayOfLastWeek = lastDayOfLastWeekOfMonth(month);
  const today = new Date()

  const canFetchRecentRoles = isSameMonth(today, month) || isSameMonth(today, lastDayOfLastWeek)

  try {
    dispatch(BinnacleActions.changeGlobalLoading(true));

    const [activities, holidays, timeBalance, recentRoles] = await Promise.all([
      getActivitiesBetweenDate(firstDayOfFirstWeek, lastDayOfLastWeek),
      getHolidaysBetweenDate(firstDayOfFirstWeek, lastDayOfLastWeek),
      getTimeBalanceBetweenDate(startOfMonth(month), endOfMonth(month)),
      canFetchRecentRoles ? getRecentRoles() : undefined
    ]);

    dispatch(
      BinnacleActions.saveBinnacleData(
        month,
        activities,
        holidays,
        timeBalance[buildTimeBalanceKey(month)],
        recentRoles
      )
    );
  } catch (error) {
    dispatch(BinnacleActions.fetchGlobalFailed(error));
    throw error;
  }
};

export const fetchTimeBalanceByYear = async (
  month: Date,
  dispatch: React.Dispatch<TBinnacleActions>
) => {
  try {
    dispatch(BinnacleActions.changeLoadingTimeBalance(true));

    const endDate = isThisMonth(month) ? subDays(month, 1) : endOfMonth(month)
    const response = await getTimeBalanceBetweenDate(
      startOfYear(month),
      endDate
    );

    const amountOfMinutes = Object.values(response).reduce(
      (prevValue, currentValue) => ({
        timeWorked: prevValue.timeWorked + currentValue.timeWorked,
        timeToWork: prevValue.timeToWork + currentValue.timeToWork,
        timeDifference:
          prevValue.timeDifference + currentValue.timeDifference
      })
    );

    dispatch(BinnacleActions.updateTimeBalance(amountOfMinutes, true));
  } catch (error) {
    dispatch(BinnacleActions.changeLoadingTimeBalance(false));
    throw error;
  }
};

export const fetchTimeBalanceByMonth = async (
  month: Date,
  dispatch: React.Dispatch<TBinnacleActions>
) => {
  try {
    dispatch(BinnacleActions.changeLoadingTimeBalance(true));

    const lastValidDate = !isSameMonth(new Date(), month)
      ? endOfMonth(month)
      : new Date();

    const response = await getTimeBalanceBetweenDate(
      startOfMonth(month),
      lastValidDate
    );

    dispatch(BinnacleActions.changeLoadingTimeBalance(false));

    const key = buildTimeBalanceKey(month)

    dispatch(
      BinnacleActions.updateTimeBalance(
        {
          timeWorked: response[key].timeWorked,
          timeToWork: response[key].timeToWork,
          timeDifference:
            response[key].timeDifference
        },
        false
      )
    );
  } catch (error) {
    dispatch(BinnacleActions.changeLoadingTimeBalance(false));
    throw error;
  }
};
