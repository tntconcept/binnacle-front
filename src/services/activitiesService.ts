import { axiosClient } from "services/axiosClient";
import { ACTIVITIES_ENDPOINT } from "services/endpoints";
import { formatDateForRequest } from "utils/calendarUtils";
import axios from "axios";
import { parseISO } from "date-fns";

export interface IActivity {
  id: number;
  startDate: Date;
  duration: number;
  description: string;
  projectRole: {
    id: number;
    name: string;
  };
  userId: number;
  billable: boolean;
  organization: {
    id: number;
    name: string;
  };
  project: {
    id: number;
    name: string;
    open: boolean;
    billable: boolean;
  };
}

export interface IActivityDay {
  date: Date;
  workedMinutes: number;
  activities: IActivity[];
}

export const getActivitiesBetweenDate = async (
  startDate: Date,
  endDate: Date
) => {
  return await axiosClient.get<IActivityDay[]>(ACTIVITIES_ENDPOINT, {
    params: {
      startDate: formatDateForRequest(startDate),
      endDate: formatDateForRequest(endDate)
    },
    // @ts-ignore
    transformResponse: axios.defaults.transformResponse.concat(
      (data: IActivityDay[]) => {
        return data.map(activityResponse => ({
          date: parseISO((activityResponse.date as unknown) as string),
          workedMinutes: activityResponse.workedMinutes,
          activities: activityResponse.activities.map(activity => ({
            ...activity,
            startDate: parseISO((activity.startDate as unknown) as string)
          }))
        }));
      }
    )
  });
};
