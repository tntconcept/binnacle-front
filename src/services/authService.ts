import { axiosClient } from "services/axiosClient";
import { AUTH_ENDPOINT, USER_ENDPOINT } from "services/endpoints";

export interface IResponseToken {
  access_token: string;
  token_type: "bearer";
  refresh_token: string;
  expires_in: number;
  scope: string;
  jti: string;
}

export interface IUser {
  id: number;
  username: string;
  name: string;
  departmentId: number;
  email?: string;
  genre: string;
  hiringDate: string; // Todo maybe i need to change to date and map the string to date
  photoUrl?: string;
  role: {
    id: number;
    name: string;
  };
  dayDuration?: number;
}

const authCredentials = {
  username: "tnt-client",
  password: "Client-TNT-v1"
};

export const getOAuthToken = async (username: string, password: string) =>
  await axiosClient.post<IResponseToken>(AUTH_ENDPOINT, undefined, {
    params: {
      grant_type: "password",
      username: username,
      password: password
    },
    auth: authCredentials
  });

/*export const refreshOAuthToken = async (refreshToken: string) =>
  await axiosClient.request<IResponseToken>({
    method: "post",
    url: AUTH_ENDPOINT,
    params: {
      grant_type: "refresh_token",
      refresh_token: refreshToken
    },
    auth: authCredentials
  });*/

export const refreshOAuthToken = async (refreshToken: string) =>
  await axiosClient.post<IResponseToken>(AUTH_ENDPOINT, undefined, {
    params: {
      grant_type: "refresh_token",
      refresh_token: refreshToken
    },
    auth: authCredentials
  });

export const getLoggedUser = async () =>
  await axiosClient.get<IUser>(USER_ENDPOINT);
