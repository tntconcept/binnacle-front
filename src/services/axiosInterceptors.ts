// Code from https://github.com/Flyrell/axios-auth-refresh
/** @type {Object} */
import { AxiosError, AxiosInstance } from "axios";
import { axiosClient } from "services/axiosClient";
import { getToken, saveToken } from "core/contexts/AuthContext/tokenUtils";
import { refreshOAuthToken } from "services/authService";

const defaults = {
  /** @type {Number[]} */
  statusCodes: [
    401 // Unauthorized
  ]
};

let _refreshCall: any = null;

/**
 * Creates an authentication refresh interceptor that binds to any error response.
 * If the response code is 401, interceptor tries to call the refreshTokenCall which must return a Promise.
 * While refreshTokenCall is running, all new requests are intercepted and waiting for it to resolve.
 * After Promise is resolved/rejected the authentication refresh interceptor is revoked.
 * @param {AxiosInstance|Function} axios - axios instance
 * @param {Function} refreshTokenCall - refresh token call which must return a Promise
 * @param retryCondition
 * @return {AxiosInstance}
 */
function createAuthRefreshInterceptor(
  axios: AxiosInstance,
  refreshTokenCall: (error: AxiosError) => Promise<any>,
  retryCondition?: () => boolean | undefined
) {
  const id = axios.interceptors.response.use(
    res => res,
    error => {
      // Reject promise if the error status is not in options.ports or defaults.ports
      const statusCodes = defaults.statusCodes;

      if (
        !error.response ||
        (error.response.status &&
          statusCodes!.indexOf(+error.response.status) === -1)
      ) {
        return Promise.reject(error);
      }

      // Reject promise if the custom condition is not satisfied (default: always retry)
      const retry =
        typeof retryCondition === "function" ? retryCondition() : true;

      if (!retry) {
        return Promise.reject(error);
      }

      // Remove the interceptor to prevent a loop
      // in case token refresh also causes the 401
      axios.interceptors.response.eject(id);

      const refreshCall = _refreshCall ? _refreshCall : refreshTokenCall(error);

      // Create interceptor that will bind all the others requests
      // until refreshTokenCall is resolved
      const requestQueueInterceptorId = axios.interceptors.request.use(
        request => refreshCall.then(() => request)
      );

      // When response code is 401 (Unauthorized), try to refresh the token.
      return refreshCall
        .then(() => {
          axios.interceptors.request.eject(requestQueueInterceptorId);
          return axios(error.response.config);
        })
        .catch((error: any) => {
          axios.interceptors.request.eject(requestQueueInterceptorId);
          return Promise.reject(error);
        })
        .finally(() => {
          _refreshCall = null;
          createAuthRefreshInterceptor(axios, refreshTokenCall);
        });
    }
  );
  return axios;
}

// Function that will be called to determine whether to attempt refresh
const retryAuthCondition = () => !!getToken("access_token");

const refreshAuthLogic = () => {
  const refreshToken = getToken("refresh_token");
  return refreshOAuthToken(refreshToken!).then(response => {
    saveToken(response.data.access_token, "access_token");
    return Promise.resolve();
  });
};

// Instantiate the interceptor (you can chain it as it returns the axios instance)
createAuthRefreshInterceptor(axiosClient, refreshAuthLogic, retryAuthCondition);
