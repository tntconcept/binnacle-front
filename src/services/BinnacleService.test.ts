import {fetchBinnacleData, fetchTimeBalanceByMonth, fetchTimeBalanceByYear} from "services/BinnacleService"
import fetchMock from 'fetch-mock/es5/client'
import {IActivityDay} from "api/interfaces/IActivity"
import {IHolidaysResponse} from "api/interfaces/IHolidays"
import {ITimeBalanceResponse} from "api/interfaces/ITimeBalance"
import endpoints from "api/endpoints"

describe("Binnacle Service", () => {
  const month = new Date(2019, 0, 10);

  const activitiesDate: IActivityDay[] = [
    {
      activities: [],
      workedMinutes: 0,
      date: new Date(2019, 1, 20)
    }
  ];

  const holidaysResponse: IHolidaysResponse = {
    publicHolidays: [],
    privateHolidays: []
  };

  const timeBalance: ITimeBalanceResponse = {
    "1": {
      timeDifference: -100,
      timeToWork: 100,
      timeWorked: 0
    }
  };

  afterEach(fetchMock.reset)

  it("should fetch all binnacle data correctly", async () => {
    fetchMock
      .getOnce("path:/" + endpoints.activities, activitiesDate)
      .getOnce("path:/" + endpoints.holidays, holidaysResponse)
      .getOnce("path:/" + endpoints.timeBalance, timeBalance)
      .getOnce("path:/" + endpoints.recentProjectRoles, []);

    const dispatch = jest.fn();
    const isTimeCalculatedByYear = false;

    await fetchBinnacleData(month, isTimeCalculatedByYear, dispatch);

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        loadingData: true
      })
    );
    expect(dispatch).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        activities: activitiesDate,
        holidays: holidaysResponse,
        timeBalance: timeBalance["1"],
        recentRoles: []
      })
    );

  });

  it("should dispatch fetch failed when any request fails", async () => {
    fetchMock
      .getOnce("path:/" + endpoints.activities, activitiesDate)
      .getOnce("path:/" + endpoints.holidays, 400)

    const dispatch = jest.fn();
    const isTimeCalculatedByYear = false;

    await expect(
      fetchBinnacleData(month, isTimeCalculatedByYear, dispatch)
    ).rejects.toMatchInlineSnapshot(`[HTTPError: Bad Request]`);

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(dispatch).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        loadingData: true
      })
    );
    expect(dispatch).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        error: expect.any(Error)
      })
    );
  });

  it("should fetch time balance by YEAR", async () => {
    fetchMock.get("path:/" + endpoints.timeBalance, timeBalance)

    const dispatch = jest.fn();

    await fetchTimeBalanceByYear(month, dispatch);

    // Calls two times change loading state of time balance
    expect(dispatch).toHaveBeenCalledTimes(2);

    expect(dispatch).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        loadingTimeBalance: true
      })
    );
    expect(dispatch).toHaveBeenNthCalledWith(
      2,
      expect.objectContaining({
        isCalculatedByYear: true,
        timeBalance: {
          timeWorked: timeBalance[1].timeWorked,
          timeToWork: timeBalance[1].timeToWork,
          differenceInMinutes: -100
        }
      })
    );
  });

  it("should throw error when fetching time by YEAR fails", async () => {
    fetchMock.getOnce("path:/" + endpoints.timeBalance, 400);

    const dispatch = jest.fn();

    await expect(
      fetchTimeBalanceByYear(month, dispatch)
    ).rejects.toMatchInlineSnapshot(`[HTTPError: Bad Request]`);

    // Calls two times change loading state of time balance
    expect(dispatch).toHaveBeenCalledTimes(2);
  });

  it("should fetch time balance by MONTH", async () => {
    fetchMock.getOnce("path:/" + endpoints.timeBalance, timeBalance);

    const dispatch = jest.fn();

    await fetchTimeBalanceByMonth(month, dispatch);

    // Calls two times change loading state of time balance
    expect(dispatch).toHaveBeenCalledTimes(3);

    expect(dispatch).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        loadingTimeBalance: true
      })
    );
    expect(dispatch).toHaveBeenNthCalledWith(
      3,
      expect.objectContaining({
        isCalculatedByYear: false,
        timeBalance: {
          timeWorked: timeBalance[1].timeWorked,
          timeToWork: timeBalance[1].timeToWork,
          differenceInMinutes: -100
        }
      })
    );
  });

  it("should throw error when fetching time by MONTH fails", async () => {
    fetchMock.getOnce("path:/" + endpoints.timeBalance, 400);

    const dispatch = jest.fn();

    await expect(
      fetchTimeBalanceByMonth(month, dispatch)
    ).rejects.toMatchInlineSnapshot(`[HTTPError: Bad Request]`);

    // Calls two times change loading state of time balance
    expect(dispatch).toHaveBeenCalledTimes(2);
  });
});
