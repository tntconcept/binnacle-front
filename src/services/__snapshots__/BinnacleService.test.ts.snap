// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Binnacle Service should dispatch fetch failed when any request fails 2`] = `
Array [
  Array [
    Object {
      "loadingData": true,
      "type": "CHANGE_GLOBAL_LOADING",
    },
  ],
  Array [
    Object {
      "error": [HTTPError: Bad Request],
      "type": "FETCH_GLOBAL_FAILED",
    },
  ],
]
`;

exports[`Binnacle Service should fetch all binnacle data correctly 1`] = `
Array [
  Array [
    Object {
      "loadingData": true,
      "type": "CHANGE_GLOBAL_LOADING",
    },
  ],
  Array [
    Object {
      "activities": Array [
        Object {
          "activities": Array [],
          "date": 2019-02-19T23:00:00.000Z,
          "workedMinutes": 0,
        },
      ],
      "holidays": Object {
        "privateHolidays": Array [],
        "publicHolidays": Array [],
      },
      "month": 2019-01-09T23:00:00.000Z,
      "recentRoles": Array [],
      "timeBalance": Object {
        "timeDifference": -100,
        "timeToWork": 100,
        "timeWorked": 0,
      },
      "type": "SAVE_BINNACLE_DATA",
    },
  ],
]
`;

exports[`Binnacle Service should fetch time balance by MONTH 1`] = `
Array [
  Array [
    Object {
      "loadingTimeBalance": true,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
  Array [
    Object {
      "loadingTimeBalance": false,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
  Array [
    Object {
      "isCalculatedByYear": false,
      "timeBalance": Object {
        "timeDifference": -100,
        "timeToWork": 100,
        "timeWorked": 0,
      },
      "type": "UPDATE_TIME_BALANCE",
    },
  ],
]
`;

exports[`Binnacle Service should fetch time balance by YEAR 1`] = `
Array [
  Array [
    Object {
      "loadingTimeBalance": true,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
  Array [
    Object {
      "isCalculatedByYear": true,
      "timeBalance": Object {
        "timeDifference": -100,
        "timeToWork": 100,
        "timeWorked": 0,
      },
      "type": "UPDATE_TIME_BALANCE",
    },
  ],
]
`;

exports[`Binnacle Service should throw error when fetching time by MONTH fails 2`] = `
Array [
  Array [
    Object {
      "loadingTimeBalance": true,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
  Array [
    Object {
      "loadingTimeBalance": false,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
]
`;

exports[`Binnacle Service should throw error when fetching time by YEAR fails 2`] = `
Array [
  Array [
    Object {
      "loadingTimeBalance": true,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
  Array [
    Object {
      "loadingTimeBalance": false,
      "type": "CHANGE_LOADING_TIME_BALANCE",
    },
  ],
]
`;
